File: backend/src/services/indicator-calculators/evwma-calculator.ts
// src/services/indicator-calculators/evwma-calculator.ts
// Translated from: services/indicator_calculators/evwma_calculator.py

import type { OHLCVBar } from '../../core/types';

export interface EvwmaResult {
  evwma: (number | null)[];
}

/**
 * Calculate EVWMA (Exponential Volume Weighted Moving Average).
 *
 * EVWMA combines volume-weighted price with exponential smoothing:
 * 1. Calculate typical price (HLC/3)
 * 2. Calculate volume-weighted price (typical_price * volume)
 * 3. Use either cumulative volume or rolling window volume
 * 4. Calculate VWMA = volume_weighted_price_sum / volume_sum
 * 5. Apply exponential smoothing with alpha = 2 / (length + 1)
 *
 * @param bars - List of OHLCV bars
 * @param length - Period for exponential smoothing
 * @param useCumVolume - If true, use cumulative volume; if false, use rolling window
 * @param rollWindow - Window size for rolling volume (required if useCumVolume=false)
 * @returns Dictionary with 'evwma' as a list of calculated values for each bar.
 */
export function calculateEvwma(
  bars: readonly OHLCVBar[],
  length: number,
  useCumVolume: boolean = false,
  rollWindow: number | null = null
): EvwmaResult {
  if (length <= 0) {
    return { evwma: new Array(bars.length).fill(null) };
  }

  if (!useCumVolume && rollWindow === null) {
    throw new Error('rollWindow must be provided when useCumVolume=false');
  }

  if (rollWindow === null) {
    rollWindow = length;
  }

  if (!bars || bars.length === 0) {
    return { evwma: [] };
  }

  const results: (number | null)[] = new Array(bars.length).fill(null);

  // Calculate typical price and volume-weighted price for each bar
  const typicalPrices: number[] = [];
  const vwpValues: number[] = []; // volume-weighted price
  const volumes: number[] = [];

  for (const bar of bars) {
    const typicalPrice = (bar.high + bar.low + bar.close) / 3.0;
    const volume = bar.volume ?? 0.0;
    typicalPrices.push(typicalPrice);
    vwpValues.push(typicalPrice * volume);
    volumes.push(volume);
  }

  // Calculate VWMA (volume-weighted moving average) for each bar
  const vwmaValues: (number | null)[] = [];

  if (useCumVolume) {
    // Use cumulative volume
    let cumVwp = 0.0;
    let cumVolume = 0.0;

    for (let i = 0; i < bars.length; i++) {
      cumVwp += vwpValues[i];
      cumVolume += volumes[i];

      if (cumVolume > 0) {
        vwmaValues.push(cumVwp / cumVolume);
      } else {
        vwmaValues.push(null);
      }
    }
  } else {
    // Use rolling window volume
    for (let i = 0; i < bars.length; i++) {
      const windowStart = Math.max(0, i - rollWindow! + 1);
      const windowVwp = vwpValues.slice(windowStart, i + 1).reduce((a, b) => a + b, 0);
      const windowVolume = volumes.slice(windowStart, i + 1).reduce((a, b) => a + b, 0);

      if (windowVolume > 0) {
        vwmaValues.push(windowVwp / windowVolume);
      } else {
        vwmaValues.push(null);
      }
    }
  }

  // Apply exponential smoothing to VWMA values
  const alpha = 2.0 / (length + 1);
  let evwma: number | null = null;

  for (let i = 0; i < bars.length; i++) {
    const vwma = vwmaValues[i];

    if (vwma === null) {
      evwma = null;
      continue;
    }

    if (evwma === null) {
      // Try to find a valid window to initialize EVWMA
      const windowStart = Math.max(0, i - length + 1);
      const window = vwmaValues.slice(windowStart, i + 1);
      const validValues = window.filter((v): v is number => v !== null);

      if (validValues.length >= length) {
        // Initialize with simple average of first length values
        evwma = validValues.slice(0, length).reduce((a, b) => a + b, 0) / length;
      } else {
        continue;
      }
    }

    // Apply exponential smoothing
    evwma = vwma * alpha + evwma * (1 - alpha);

    // Only set result if we have enough data
    if (i >= length - 1) {
      results[i] = evwma;
    }
  }

  return { evwma: results };
}

/**
 * Calculate rolling correlation between two series.
 *
 * @param x - First series (can contain null values)
 * @param y - Second series (can contain null values)
 * @param window - Window size for rolling correlation
 * @returns List of correlation values, with null for insufficient data
 */
export function calculateRollingCorrelation(
  x: (number | null)[],
  y: (number | null)[],
  window: number
): (number | null)[] {
  if (x.length !== y.length) {
    throw new Error('Series x and y must have the same length');
  }

  if (window <= 0) {
    return new Array(x.length).fill(null);
  }

  const results: (number | null)[] = new Array(x.length).fill(null);

  for (let i = window - 1; i < x.length; i++) {
    const windowX = x.slice(i - window + 1, i + 1);
    const windowY = y.slice(i - window + 1, i + 1);

    // Filter out null values
    const validPairs: [number, number][] = [];
    for (let j = 0; j < windowX.length; j++) {
      const a = windowX[j];
      const b = windowY[j];
      if (a !== null && b !== null) {
        validPairs.push([a, b]);
      }
    }

    if (validPairs.length < 2) {
      results[i] = null;
      continue;
    }

    const validX = validPairs.map((pair) => pair[0]);
    const validY = validPairs.map((pair) => pair[1]);

    // Calculate mean
    const meanX = validX.reduce((a, b) => a + b, 0) / validX.length;
    const meanY = validY.reduce((a, b) => a + b, 0) / validY.length;

    // Calculate covariance and variances
    let covariance = 0;
    let varianceX = 0;
    let varianceY = 0;

    for (let j = 0; j < validX.length; j++) {
      covariance += (validX[j] - meanX) * (validY[j] - meanY);
      varianceX += (validX[j] - meanX) ** 2;
      varianceY += (validY[j] - meanY) ** 2;
    }

    // Calculate correlation
    const denominator = Math.sqrt(varianceX * varianceY);
    if (denominator === 0) {
      results[i] = null;
    } else {
      const correlation = covariance / denominator;
      results[i] = correlation;
    }
  }

  return results;
}

File: legacy/services/indicator_calculators/evwma_calculator.py
from collections.abc import Sequence
from typing import Any

from core.types_registry import OHLCVBar


def calculate_evwma(
    bars: Sequence[OHLCVBar],
    length: int,
    use_cum_volume: bool = False,
    roll_window: int | None = None,
) -> dict[str, Any]:
    """
    Calculate EVWMA (Exponential Volume Weighted Moving Average).

    EVWMA combines volume-weighted price with exponential smoothing:
    1. Calculate typical price (HLC/3)
    2. Calculate volume-weighted price (typical_price * volume)
    3. Use either cumulative volume or rolling window volume
    4. Calculate VWMA = volume_weighted_price_sum / volume_sum
    5. Apply exponential smoothing with alpha = 2 / (length + 1)

    Args:
        bars: List of OHLCV bars
        length: Period for exponential smoothing
        use_cum_volume: If True, use cumulative volume; if False, use rolling window
        roll_window: Window size for rolling volume (required if use_cum_volume=False)

    Returns:
        Dictionary with 'evwma' as a list of calculated values for each bar.
        Returns None for bars before sufficient data is available.
    """
    if length <= 0:
        return {"evwma": [None] * len(bars)}

    if not use_cum_volume and roll_window is None:
        raise ValueError("roll_window must be provided when use_cum_volume=False")

    if roll_window is None:
        roll_window = length

    if not bars or len(bars) == 0:
        return {"evwma": []}

    results: list[float | None] = [None] * len(bars)

    # Calculate typical price and volume-weighted price for each bar
    typical_prices: list[float] = []
    vwp_values: list[float] = []  # volume-weighted price
    volumes: list[float] = []

    for bar in bars:
        typical_price = (bar["high"] + bar["low"] + bar["close"]) / 3.0
        volume = bar.get("volume", 0.0) or 0.0
        typical_prices.append(typical_price)
        vwp_values.append(typical_price * volume)
        volumes.append(volume)

    # Calculate VWMA (volume-weighted moving average) for each bar
    vwma_values: list[float | None] = []

    if use_cum_volume:
        # Use cumulative volume
        cum_vwp = 0.0
        cum_volume = 0.0

        for i in range(len(bars)):
            cum_vwp += vwp_values[i]
            cum_volume += volumes[i]

            if cum_volume > 0:
                vwma_values.append(cum_vwp / cum_volume)
            else:
                vwma_values.append(None)
    else:
        # Use rolling window volume
        for i in range(len(bars)):
            window_start = max(0, i - roll_window + 1)
            window_vwp = sum(vwp_values[window_start : i + 1])
            window_volume = sum(volumes[window_start : i + 1])

            if window_volume > 0:
                vwma_values.append(window_vwp / window_volume)
            else:
                vwma_values.append(None)

    # Apply exponential smoothing to VWMA values
    alpha = 2.0 / (length + 1)
    evwma: float | None = None

    for i in range(len(bars)):
        vwma = vwma_values[i]

        if vwma is None:
            evwma = None
            continue

        if evwma is None:
            # Try to find a valid window to initialize EVWMA
            window_start = max(0, i - length + 1)
            window = vwma_values[window_start : i + 1]
            valid_values = [v for v in window if v is not None]

            if len(valid_values) >= length:
                # Initialize with simple average of first length values
                evwma = sum(valid_values[:length]) / length
            else:
                continue

        # Apply exponential smoothing
        evwma = vwma * alpha + evwma * (1 - alpha)

        # Only set result if we have enough data
        if i >= length - 1:
            results[i] = evwma

    return {"evwma": results}


def calculate_rolling_correlation(
    x: list[float | None], y: list[float | None], window: int
) -> list[float | None]:
    """
    Calculate rolling correlation between two series.

    Args:
        x: First series (can contain None values)
        y: Second series (can contain None values)
        window: Window size for rolling correlation

    Returns:
        List of correlation values, with None for insufficient data
    """
    if len(x) != len(y):
        raise ValueError("Series x and y must have the same length")

    if window <= 0:
        return [None] * len(x)

    results: list[float | None] = [None] * len(x)

    for i in range(window - 1, len(x)):
        window_x = x[i - window + 1 : i + 1]
        window_y = y[i - window + 1 : i + 1]

        # Filter out None values
        valid_pairs = [(a, b) for a, b in zip(window_x, window_y) if a is not None and b is not None]

        if len(valid_pairs) < 2:
            results[i] = None
            continue

        valid_x = [pair[0] for pair in valid_pairs]
        valid_y = [pair[1] for pair in valid_pairs]

        # Calculate mean
        mean_x = sum(valid_x) / len(valid_x)
        mean_y = sum(valid_y) / len(valid_y)

        # Calculate covariance and variances
        covariance = sum((valid_x[j] - mean_x) * (valid_y[j] - mean_y) for j in range(len(valid_x)))
        variance_x = sum((val - mean_x) ** 2 for val in valid_x)
        variance_y = sum((val - mean_y) ** 2 for val in valid_y)

        # Calculate correlation
        denominator = (variance_x * variance_y) ** 0.5
        if denominator == 0:
            results[i] = None
        else:
            correlation = covariance / denominator
            results[i] = correlation

    return results



